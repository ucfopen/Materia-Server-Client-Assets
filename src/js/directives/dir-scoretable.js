// TODO: This file was created by bulk-decaffeinate.
// Sanity-check the conversion and remove this comment.
/*
 * decaffeinate suggestions:
 * DS101: Remove unnecessary use of Array.from
 * DS102: Remove unnecessary code created because of implicit returns
 * DS205: Consider reworking code to avoid use of IIFEs
 * DS207: Consider shorter variations of null checks
 * Full docs: https://github.com/decaffeinate/decaffeinate/blob/master/docs/suggestions.md
 */
'use strict'

const app = angular.module('materia')
app.directive('scoreTable', function(selectedWidgetSrv, $window) {
	return {
		restrict: 'A',
		link($scope, $element, $attrs) {
			const widgetId = selectedWidgetSrv.getSelectedId()
			const tableSort = 'desc'
			const userCount = []
			const users = {}
			const { year } = $attrs
			let { term } = $attrs
			let masterUserList = {}
			$scope.users = {}
			$scope.selectedUser = null

			const logs = selectedWidgetSrv.getPlayLogsForSemester(term, year)
			logs.then(function(data) {
				// process play logs into records for each user
				angular.forEach(data, function(log, index) {
					const uid = log.user_id
					const name = log.last ? `${log.last}, ${log.first}` : 'Guests'

					if (users[uid] == null) {
						users[uid] = {
							uid,
							name,
							scores: {}
						}
					}

					// make the score percentage readable
					let percent = 0
					if (log.done === '1') {
						percent = parseFloat(log.perc)
							.toFixed(2)
							.replace('.00', '')
					}

					// make the play duration readable
					let duration = 0
					const mins = (log.elapsed - log.elapsed % 60) / 60
					const secs = log.elapsed % 60

					if (mins !== 0) {
						duration = `${mins}m ${secs}s`
					} else {
						duration = `${secs}s`
					}

					return (users[uid].scores[log.time.toString()] = {
						date: new Date(log.time * 1000).toDateString(),
						percent,
						elapsed: duration,
						complete: log.done,
						id: log.id
					})
				})

				masterUserList = $scope.users = users
			})

			$scope.setSelectedUser = id => {
				$scope.selectedUser = $scope.users[id]
			}

			$scope.showScorePage = function(scoreId) {
				$window.open(`${BASE_URL}scores/${widgetId}/#single-${scoreId}`)
				return true
			}

			$scope.searchStudentActivity = function(query) {
				if (query === '') {
					$scope.users = masterUserList
					return
				}

				$scope.selectedUser = null
				const sanitized = query.toLowerCase().replace(/,/g, ' ')
				const hits = {}
				const terms = sanitized.split(' ')

				// loop over mast users to check if any search word matches the user name
				angular.forEach(masterUserList, user => {
					let name = user.name.toLowerCase()
					terms.forEach(term => {
						if (name.includes(term)) {
							hits[user.uid] = user
						}
					})
				})

				$scope.users = hits
			}
		}
	}
})
