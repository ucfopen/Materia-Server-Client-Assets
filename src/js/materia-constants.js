const app = angular.module('materia');

app.constant('PLAYER', {
	LOG_INTERVAL: 10000, // How often to send logs to the server
	RETRY_LIMIT: 15, // When the logs fail to send, retry how many times before switching to slow mode?
	RETRY_FAST: 1000,
	RETRY_SLOW: 10000,
	EMBED_TARGET: 'container'
}
); // id of the container to put the flash in

app.constant('OBJECT_TYPES', {
	QUESTION: 1,
	ASSET: 2,
	WIDGET: 3,
	WIDGET_INSTANCE: 4
}
);

app.constant('ACCESS', {
	VISIBLE: 1,
	PLAY: 5,
	SCORE: 10,
	DATA: 15,
	EDIT: 20,
	COPY: 25,
	FULL: 30,
	SHARE: 35
})
