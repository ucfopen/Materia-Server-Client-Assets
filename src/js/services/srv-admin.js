const app = angular.module('materia')
app.service('adminSrv', function() {
	const getWidgets = callback => {
		Materia.Coms.Json.get('/api/admin/widgets', callback)
	}

	const saveWidget = (widget, callback) => {
		Materia.Coms.Json.post(`/api/admin/widget/${widget.id}`, widget, callback)
	}

	const searchUsers = (str, callback) => {
		Materia.Coms.Json.get(`/api/admin/user_search/${str}`, callback)
	}

	const lookupUser = (userId, callback) => {
		Materia.Coms.Json.get(`/api/admin/user/${userId}`, callback)
	}

	const saveUser = (obj, callback) => {
		Materia.Coms.Json.post(`/api/admin/user/${obj.id}`, obj, callback)
	}

	return {
		getWidgets,
		saveWidget,
		searchUsers,
		lookupUser,
		saveUser
	}
})
