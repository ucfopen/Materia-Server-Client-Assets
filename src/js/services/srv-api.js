const app = angular.module('materia')
app.service('apiServ', () => {
	const gatewayURL = API_LINK

	const filterError = data => {
		if (data != null && data.msg != null && data.title != null && data.type != null) {
			showErorr(data)
		}
	}

	var showErorr = data => {
		if (data.title === 'Invalid Login') {
			// redirect to login page
			window.location = BASE_URL + 'login'
		}
	}

	// public methods
	return {
		showErorr,
		filterError
	}
})
