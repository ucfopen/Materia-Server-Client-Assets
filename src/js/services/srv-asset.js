const app = angular.module('materia')
app.service('AssetSrv', function () {
	const deleteAsset = (asset) =>
		Materia.Coms.Json.del(`/api/asset/delete/${encodeURIComponent(asset.id)}`)
	const restoreAsset = (asset) =>
		Materia.Coms.Json.post(`/api/asset/restore/${encodeURIComponent(asset.id)}`)

	return {
		deleteAsset,
		restoreAsset
	}
})
