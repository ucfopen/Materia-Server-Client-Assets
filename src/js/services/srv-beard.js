// TODO: This file was created by bulk-decaffeinate.
// Sanity-check the conversion and remove this comment.
/*
 * decaffeinate suggestions:
 * DS102: Remove unnecessary code created because of implicit returns
 * Full docs: https://github.com/decaffeinate/decaffeinate/blob/master/docs/suggestions.md
 */
// TODO: rip out redundant methods
const app = angular.module('materia');
app.service('beardServ', function() {

	const beards = ['dusty_full', 'black_chops', 'grey_gandalf', 'red_soul'];

	const getRandomBeard = () => beards[Math.floor(Math.random() * beards.length)];

	return {getRandomBeard};
});

