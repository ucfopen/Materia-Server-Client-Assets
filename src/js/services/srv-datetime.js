const app = angular.module('materia')
app.service('dateTimeServ', function() {
	const parseObjectToDateString = function(time) {
		const timeObj = new Date(time * 1000)
		const year = String(timeObj.getFullYear())
		return timeObj.getMonth() + 1 + '/' + timeObj.getDate() + '/' + year.substr(2)
	}

	const parseTime = function(time) {
		const timeObj = new Date(time * 1000)
		let amPm = 'am'
		let hour = timeObj.getHours()
		let minute = timeObj.getMinutes()

		if (minute < 10) {
			minute = `0${minute}`
		}

		if (hour > 11) {
			if (hour !== 12) {
				hour -= 12
			}
			amPm = 'pm'
		} else {
			if (hour === 0) {
				hour = '12'
			}
		}

		return `${hour}:${minute}${amPm}`
	}

	const fixTime = (time, servTime) => {
		const timeToFix = new Date(time).getTime()
		const serverDateFromPage = servTime

		// calculate client/server time difference
		const now = new Date()
		const clientUTCDate = new Date(
			now.getUTCFullYear(),
			now.getUTCMonth(),
			now.getUTCDate(),
			now.getUTCHours(),
			now.getUTCMinutes(),
			now.getUTCSeconds()
		)
		const serverUTCDate = new Date(serverDateFromPage)
		const clientUTCTimestamp = clientUTCDate.getTime()
		const serverUTCTimestamp = serverUTCDate.getTime()
		const offset = serverUTCTimestamp - clientUTCTimestamp

		const newDate = new Date(timeToFix + offset)
		const fullHour = newDate.getHours()
		const shortHour = fullHour % 12 === 0 ? 12 : fullHour % 12
		let fixedDateStr = shortHour + ':' + String(`00${newDate.getMinutes()}`).slice(-2)

		if (fullHour > 11) {
			fixedDateStr += 'pm'
		} else {
			fixedDateStr += 'am'
		}

		return fixedDateStr
	}

	return {
		parseObjectToDateString,
		parseTime,
		fixTime
	}
})
