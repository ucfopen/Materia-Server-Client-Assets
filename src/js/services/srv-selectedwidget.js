const app = angular.module('materia')
app.service('selectedWidgetSrv', function($rootScope, $q, OBJECT_TYPES) {
	const STORAGE_TABLE_MAX_ROWS_SHOWN = 100

	const selectedData = null
	const storageData = null
	const instId = null

	// Refactored variables
	let _widget = null
	let _dateRanges = null
	let _scoreData = null
	let _storageData = null

	// get and set _widget
	const set = widget => {
		_scoreData = null
		_storageData = null
		_widget = widget
		$rootScope.$broadcast('selectedWidget.update')
	}

	const get = () => _widget

	const getSelectedId = () => _widget.id

	const getScoreSummaries = () => {
		const deferred = $q.defer()

		if (_scoreData) {
			deferred.resolve(_scoreData)
		} else {
			let loadingId = _widget.id
			Materia.Coms.Json.send('score_summary_get', [loadingId, true]).then(data => {
				if (loadingId !== _widget.id) {
					return deferred.reject()
				}

				_scoreData = {
					list: [],
					map: {},
					last: undefined
				}

				if (data !== null && data.length > 0) {
					const map = {}
					data.forEach(d => {
						map[d.id] = d
					})

					_scoreData = {
						list: data,
						map,
						last: data[0]
					}
				}

				deferred.resolve(_scoreData)
			})
		}

		return deferred.promise
	}

	const getUserPermissions = () => {
		return Materia.Coms.Json.send('permissions_get', [
			OBJECT_TYPES.WIDGET_INSTANCE,
			_widget.id
		]).then(perms => {
			const permsObject = {
				user: perms.user_perms,
				widget: perms.widget_user_perms
			}

			return permsObject
		})
	}

	const getPlayLogsForSemester = (term, year) => {
		const deferred = $q.defer()

		Materia.Coms.Json.send('play_logs_get', [_widget.id, term, year]).then(logs => {
			const semesterKey = `${year}${term.toLowerCase()}`
			const logsForSemester = []
			angular.forEach(logs, (log, key) => {
				const timestamp = log.time
				const logMeta = getSemesterFromTimestamp(timestamp)
				const semesterString = logMeta.year + logMeta.semester.toLowerCase()

				if (semesterString === semesterKey) {
					logsForSemester.push(log)
				}
			})

			deferred.resolve(logsForSemester)
		})

		return deferred.promise
	}

	const getDateRanges = () => {
		const deferred = $q.defer()
		if (_dateRanges == null) {
			Materia.Coms.Json.send('semester_date_ranges_get', []).then(data => {
				_dateRanges = data
				deferred.resolve(data)
			})
		} else {
			deferred.resolve(_dateRanges)
		}
		return deferred.promise
	}

	const getSemesterFromTimestamp = timestamp => {
		return _dateRanges.find(
			r => timestamp >= parseInt(r.start, 10) && timestamp <= parseInt(r.end, 10)
		)
	}

	const getStorageData = (loadIfNotCached = true) => {
		const deferred = $q.defer()

		if (_storageData != null) {
			deferred.resolve(_storageData)
		} else if (loadIfNotCached) {
			Materia.Coms.Json.send('play_storage_get', [_widget.id]).then(data => {
				_storageData = {}

				const temp = {}

				// process semester data and organize by table name
				angular.forEach(
					data,
					(tableData, tableName) => (temp[tableName] = _processDataIntoSemesters(tableData))
				)

				// have to loop through each table present in the storage data
				angular.forEach(temp, (semesters, tableName) =>
					// have to loop through each semester contained within each table
					angular.forEach(semesters, (semesterData, semesterId) => {
						if (typeof _storageData[semesterId] === 'undefined') {
							_storageData[semesterId] = {}
						}

						if (semesterData.length > STORAGE_TABLE_MAX_ROWS_SHOWN) {
							_storageData[semesterId][tableName] = {
								truncated: true,
								total: semesterData.length,
								data: semesterData.slice(0, STORAGE_TABLE_MAX_ROWS_SHOWN)
							}
						} else {
							_storageData[semesterId][tableName] = { truncated: false, data: semesterData }
						}

						_storageData[semesterId][tableName].data = _normalizeStorageDataColumns(
							_storageData[semesterId][tableName].data
						)
					})
				)
				deferred.resolve(_storageData)
			})
		} else {
			deferred.resolve()
		}

		return deferred.promise
	}

	const _processDataIntoSemesters = logs => {
		const semesters = {}
		let timestamp = null

		angular.forEach(logs, (log, index) => {
			timestamp = log.play.time
			const logMeta = getSemesterFromTimestamp(timestamp)
			const semesterString = logMeta.year + ' ' + logMeta.semester.toLowerCase()

			if (!semesters[semesterString]) {
				semesters[semesterString] = []
			}
			semesters[semesterString].push(log)
		})

		return semesters
	}

	//  storage data doesn't really enforce a schema.
	//  this function determines every field used throughout the
	//  storage data and then applies that schema to each item.
	const _normalizeStorageDataColumns = rows => {
		//  go through all the rows and collect the fields used:
		const fields = {}
		for (var r of rows) {
			for (let j in r.data) {
				if (typeof r.data[j] === 'undefined') {
					r.j = null
				}
			}
		}

		//  now go through each row again and add in the missing fields
		for (r of rows) {
			r.data = Object.assign({}, fields, r.data)
		}

		return rows
	}

	const getMaxRows = () => STORAGE_TABLE_MAX_ROWS_SHOWN

	const notifyAccessDenied = () => {
		$rootScope.$broadcast('selectedWidget.notifyAccessDenied')
	}

	return {
		set,
		get,
		getSelectedId,
		getScoreSummaries,
		getUserPermissions,
		getPlayLogsForSemester,
		getDateRanges,
		getSemesterFromTimestamp,
		getStorageData,
		getMaxRows,
		notifyAccessDenied
	}
})
