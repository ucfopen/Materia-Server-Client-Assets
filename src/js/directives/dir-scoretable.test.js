describe('scoreTable Directive', function() {
	let $scope
	let $compile
	let $window
	let $q
	let selectedWidgetSrv
	let data = [
		{
			id: 'one',
			time: '1519272328',
			done: '1',
			perc: '99',
			elapsed: '999',
			qset_id: '7',
			user_id: '2',
			first: 'Ian',
			last: 'Turgeon',
			username: '~author'
		},
		{
			id: 'two',
			time: '1519372329',
			done: '0',
			perc: '75',
			elapsed: '12',
			qset_id: '45',
			user_id: '5',
			first: 'Corey',
			last: 'Peterson',
			username: '~author2'
		}
	]
	let expected = {
		'2': {
			name: 'Turgeon, Ian',
			scores: {
				'1519272328': {
					complete: '1',
					date: 'Wed Feb 21 2018',
					elapsed: '16m 39s',
					id: 'one',
					percent: '99'
				}
			},
			uid: '2'
		},
		'5': {
			name: 'Peterson, Corey',
			scores: {
				'1519372329': {
					complete: '0',
					date: 'Fri Feb 23 2018',
					elapsed: '12s',
					id: 'two',
					percent: 0
				}
			},
			uid: '5'
		}
	}

	beforeEach(() => {
		require('../materia-constants')
		require('../services/srv-selectedwidget')
		require('./dir-scoretable')

		inject(function(_$compile_, _$rootScope_, _selectedWidgetSrv_, _$q_, _$window_) {
			selectedWidgetSrv = _selectedWidgetSrv_
			$compile = _$compile_
			$scope = _$rootScope_.$new()
			$q = _$q_
			$window = _$window_
		})

		let deferred = $q.defer()
		jest.spyOn(selectedWidgetSrv, 'getSelectedId').mockImplementation(() => 6)
		jest
			.spyOn(selectedWidgetSrv, 'getPlayLogsForSemester')
			.mockImplementation(() => deferred.promise)

		expect($scope.selectedUser).not.toBeDefined()
		expect($scope.setSelectedUser).not.toBeDefined()
		expect($scope.showScorePage).not.toBeDefined()
		expect($scope.searchStudentActivity).not.toBeDefined()
		let html = '<div score-table data-term="TERM" data-year="2015"></div>'
		let element = angular.element(html)
		let compiled = $compile(element)($scope)
		$scope.$digest()

		deferred.resolve(data)
		$scope.$apply()
	})

	it('is initialized on the element', function() {
		expect($scope.selectedUser).toBeNull()
		expect($scope.users).toMatchObject(expected)
		expect($scope.setSelectedUser).toBeDefined()
		expect($scope.showScorePage).toBeDefined()
		expect($scope.searchStudentActivity).toBeDefined()
	})

	it('setSelectedUser sets selectedUser object as expected', function() {
		expect($scope.selectedUser).toBeNull()
		$scope.setSelectedUser(5)
		expect($scope.selectedUser).toMatchObject(expected['5'])
	})

	it('showScorePage opens the expected url', function() {
		$window.open = jest.fn()
		global.BASE_URL = 'some_url'
		$scope.showScorePage('two')
		expect($window.open).toHaveBeenLastCalledWith('some_urlscores/6/#single-two')
	})

	it('searchStudentActivity locates users', function() {
		$scope.searchStudentActivity('Ian')
		expect($scope.users).toMatchObject({ '2': expected['2'] })

		// finds neither
		$scope.searchStudentActivity('~author')
		expect($scope.users).toMatchObject({})

		$scope.searchStudentActivity('Peterson')
		expect($scope.users).toMatchObject({ '5': expected['5'] })

		// finds all
		$scope.searchStudentActivity('')
		expect($scope.users).toMatchObject(expected)
	})

	it('searchStudentActivity resets selecteUser', function() {
		$scope.setSelectedUser(5)
		expect($scope.selectedUser).toMatchObject(expected['5'])
		$scope.searchStudentActivity('Ian')
		expect($scope.users).toMatchObject({ '2': expected['2'] })
		expect($scope.selectedUser).toBeNull()
	})
})
