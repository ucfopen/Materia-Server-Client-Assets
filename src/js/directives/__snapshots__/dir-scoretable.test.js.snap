// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scoreTable Directive is initialized on the element 1`] = `
Object {
  "2": Object {
    "name": "Turgeon, Ian",
    "scores": Object {
      "1519272328": Object {
        "complete": "1",
        "date": "Thu Feb 22 2018",
        "elapsed": "16m 39s",
        "id": "one",
        "percent": "99",
      },
    },
    "uid": "2",
  },
  "5": Object {
    "name": "Peterson, Corey",
    "scores": Object {
      "1519372329": Object {
        "complete": "0",
        "date": "Fri Feb 23 2018",
        "elapsed": "12s",
        "id": "two",
        "percent": 0,
      },
    },
    "uid": "5",
  },
}
`;

exports[`scoreTable Directive searchStudentActivity locates users 1`] = `
Object {
  "2": Object {
    "name": "Turgeon, Ian",
    "scores": Object {
      "1519272328": Object {
        "complete": "1",
        "date": "Thu Feb 22 2018",
        "elapsed": "16m 39s",
        "id": "one",
        "percent": "99",
      },
    },
    "uid": "2",
  },
}
`;

exports[`scoreTable Directive searchStudentActivity locates users 2`] = `
Object {
  "5": Object {
    "name": "Peterson, Corey",
    "scores": Object {
      "1519372329": Object {
        "complete": "0",
        "date": "Fri Feb 23 2018",
        "elapsed": "12s",
        "id": "two",
        "percent": 0,
      },
    },
    "uid": "5",
  },
}
`;

exports[`scoreTable Directive searchStudentActivity locates users 3`] = `
Object {
  "2": Object {
    "name": "Turgeon, Ian",
    "scores": Object {
      "1519272328": Object {
        "complete": "1",
        "date": "Thu Feb 22 2018",
        "elapsed": "16m 39s",
        "id": "one",
        "percent": "99",
      },
    },
    "uid": "2",
  },
  "5": Object {
    "name": "Peterson, Corey",
    "scores": Object {
      "1519372329": Object {
        "complete": "0",
        "date": "Fri Feb 23 2018",
        "elapsed": "12s",
        "id": "two",
        "percent": 0,
      },
    },
    "uid": "5",
  },
}
`;

exports[`scoreTable Directive searchStudentActivity resets selecteUser 1`] = `
Object {
  "name": "Peterson, Corey",
  "scores": Object {
    "1519372329": Object {
      "complete": "0",
      "date": "Fri Feb 23 2018",
      "elapsed": "12s",
      "id": "two",
      "percent": 0,
    },
  },
  "uid": "5",
}
`;

exports[`scoreTable Directive searchStudentActivity resets selecteUser 2`] = `
Object {
  "2": Object {
    "name": "Turgeon, Ian",
    "scores": Object {
      "1519272328": Object {
        "complete": "1",
        "date": "Thu Feb 22 2018",
        "elapsed": "16m 39s",
        "id": "one",
        "percent": "99",
      },
    },
    "uid": "2",
  },
}
`;

exports[`scoreTable Directive setSelectedUser sets selectedUser object as expected 1`] = `
Object {
  "name": "Peterson, Corey",
  "scores": Object {
    "1519372329": Object {
      "complete": "0",
      "date": "Fri Feb 23 2018",
      "elapsed": "12s",
      "id": "two",
      "percent": 0,
    },
  },
  "uid": "5",
}
`;
